Source/Core/Common/WorkQueueThread.h

```diff
@@ -20,89 +22,148 @@ class WorkQueueThread
 {
 public:
   WorkQueueThread() = default;
-  WorkQueueThread(std::function<void(T)> function) { Reset(std::move(function)); }
+  WorkQueueThread(const std::string_view name, std::function<void(T)> function)
+  {
+    Reset(name, std::move(function));
+  }
   ~WorkQueueThread() { Shutdown(); }
-  void Reset(std::function<void(T)> function)
+
+  // Shuts the current work thread down (if any) and starts a new thread with the given function
+  // Note: Some consumers of this API push items to the queue before starting the thread.
+  void Reset(const std::string_view name, std::function<void(T)> function)
   {
     Shutdown();
-    m_shutdown.Clear();
-    m_cancelled.Clear();
+    std::lock_guard lg(m_lock);
+    m_thread_name = name;
+    m_shutdown = false;
     m_function = std::move(function);
     m_thread = std::thread(&WorkQueueThread::ThreadLoop, this);
   }
 
+  // Adds an item to the work queue
   template <typename... Args>
   void EmplaceItem(Args&&... args)
   {
-    if (!m_cancelled.IsSet())
-    {
-      std::lock_guard lg(m_lock);
-      m_items.emplace(std::forward<Args>(args)...);
-    }
-    m_wakeup.Set();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.emplace(std::forward<Args>(args)...);
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  void Clear()
+  // Adds an item to the work queue
+  void Push(T&& item)
   {
-    {
-      std::lock_guard lg(m_lock);
-      m_items = std::queue<T>();
-    }
-    m_wakeup.Set();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.push(std::move(item));
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  void Cancel()
+  // Adds an item to the work queue
+  void Push(const T& item)
   {
-    m_cancelled.Set();
-    Clear();
-    Shutdown();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.push(item);
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  bool IsCancelled() const { return m_cancelled.IsSet(); }
+  // Empties the queue
+  // If the worker polls IsCanceling(), it can abort it's work when Cancelling
+  void Cancel()
+  {
+    std::unique_lock lg(m_lock);
+    if (m_shutdown)
+      return;
 
-  void Shutdown()
+    m_cancelling = true;
+    m_items = std::queue<T>();
+    m_worker_cond_var.notify_one();
+  }
+
+  // Tells the worker to shut down when it's queue is empty
+  // Blocks until the worker thread exits.
+  // If cancel is true, will Cancel before before telling the worker to exit
```

Comment:
What is the point of canceling when you are shutting down anyway?

Reply:
To shutdown faster. Otherwise it's going to continue to process work items until the queue is empty. 

For some consumers of this API (currently just GameList tracker) the work items might be quite long running, and there is no consequence to aborting them.

Reply:
Ok, I'm missing something.   You `break` out of the loop?  You should be ignoring all items right?

Reply:
We empty `m_items` before notifying the condition var.

Reply:
Yeah but all three values will notify the condition variable:

```
m_worker_cond_var.wait(
            lg, [&] { return m_shutdown || m_cancelling.load() || !m_items.empty(); });
```

And with the `break` I don't see a reason for the cancel here.

Reply:
It only waits on the condition var when the queue is already empty. 

If the queue isn't empty, it's going to keep executing items until it's empty, then it will wait on the condition var, see m_shutdown is set and then exit.

Reply:
Yes but to loop back around to my comment.  What is the point of cancelling during shutdown.  You are going to break out of the loop (never have to worry about processing other elements).